tests/test_TFNetworkRecLayer.py

```diff
@@ -3598,6 +3598,31 @@ def test_reclayer_optimize_out_access_split():
     other_subnet_layers={"split": {"class": "split", "from": ["data:source"], "size_splits": [5, 8]}})
 
 
+def test_reclayer_optimize_out_slice_nd():
+  def random_start_positions(source, **kwargs):
+    import tensorflow as tf
+    enc = source(0, as_data=True, enforce_batch_major=True, auto_convert=False)
+    enc_shape = tf.shape(enc.placeholder)  # (B,T,D)
+    enc_time_dim = enc_shape[enc.time_dim_axis]
+    return tf.random.uniform(enc_shape[:-1], 0, enc_time_dim-2, dtype=tf.dtypes.int32)  #[B,T]
+
+  check_reclayer_optimize_out(
+    {"class": "linear", "activation": None, "from": "my_layer"},
+    from_="position",
+    other_subnet_layers={
+      "my_layer": {"class": "gather_nd", "from": "base:data", "position": ":i"},
+      "window": {"class": "slice_nd2",    # no_opt: [B,4,D], opt: [B,T,4,D]
+                 "from": "base:data", "start": "data:source", "size": 4, "is_output_layer": True},
```

Comment:
> I am not sure what **input to expect** when the `slice_nd2` layer gets **optimized-out**.

I don't really understand how this can not be clear? This should be totally obvious/straightforward or not?

`data:source` is the rec layer input. Inside the loop, it is [B]. Outside the loop, it is [T,B]. (It is `position` here.)

`base:data` is independent from the loop / rec layer, so obviously nothing changes.

> It is related to #550.

I don't really understand. #550 is not even a bug but was just a question.

> I would suggest adding an option to control the behaviour of the layers (e.g. the ones comming from base network) when being pulled of `rec`.

I don't understand. What behavior are you speaking about? There is only one straightforward behavior.

Reply:
> `data:source` is the rec layer input. Inside the loop, it is [B]. Outside the loop, it is [T,B]. (It is position **here.)**

This is clear.

> `base:data` is independent from the loop / rec layer, so obviously nothing changes.

Exactly. But, instead of `base:data`,  the input could be loop-dependent and we would get an extra time axis when pulled out. (first case above)
So, I am saying that depending on whether the input comes from inside the loop or not we may get 2 different inputs and `SliceNdLayer` has to handle both properly. 

#550 was indeed a question. 
My point is very simple, sorry that I am not being clear. **If you pull out a layer from the loop, you would expect its inputs to be accumulated too(extra time dim), no matter if they are loop-dependent or not.** Or at least have a parameter to force it.

Reply:
> > `base:data` is independent from the loop / rec layer, so obviously nothing changes.
> 
> Exactly. But, instead of `base:data`, the input could be loop-dependent and we would get an extra time axis when pulled out. (first case above)

No. The `base:` prefix means it comes from the base, so it is always independently. This is independent how it is optimized (in loop or without).

Reply:
Is there any way how to make it dependent? For example copy it into an intermediate layer or so?

Otherwise, I don't know how making it work for all cases is possible. Maybe making sure that the `input` has one more time axis than `start` or `tf.tile` it otherwise?

Reply:
Why would you want that? It doesn't really make sense. Also you do not have this case. Why do you want to cover cases which you do not need? Keep it as simple as possible.